File: frontend/src/lib/sheets/pagination.ts
import type { PaginatedResponse, Pagination } from '../../types/api';

export const paginateItems = <T>(
  items: T[],
  page = 1,
  limit = 20,
): PaginatedResponse<T> => {
  const safePage = Math.max(1, page);
  const safeLimit = Math.max(1, Math.min(100, limit));
  const totalItems = items.length;
  const totalPages = Math.max(1, Math.ceil(totalItems / safeLimit));
  const start = (safePage - 1) * safeLimit;

  const pagination: Pagination = {
    page: safePage,
    limit: safeLimit,
    totalItems,
    totalPages,
  };

  return {
    items: items.slice(start, start + safeLimit),
    pagination,
  };
};

File: mobile/lib/utils/pagination.dart
import '../models/api_models.dart';

PaginatedResponse<T> paginateItems<T>(
  List<T> items, {
  int page = 1,
  int limit = 20,
}) {
  final safePage = page < 1 ? 1 : page;
  final safeLimit = limit < 1 ? 1 : (limit > 100 ? 100 : limit);
  final totalItems = items.length;
  final totalPages =
      totalItems == 0 ? 1 : (totalItems / safeLimit).ceil().clamp(1, 1 << 31);
  final start = (safePage - 1) * safeLimit;

  return PaginatedResponse(
    items: items.skip(start).take(safeLimit).toList(),
    pagination: Pagination(
      page: safePage,
      limit: safeLimit,
      totalItems: totalItems,
      totalPages: totalPages,
    ),
  );
}
